### ISSUE TITLE
User id 999 is `systemd-journal` on Debian by default.

### ISSUE THREAD

**Issue by googollee** [photoview/photoview#1009]
The docker image of photoview creates [user `photoview` with id 999](https://github.com/photoview/photoview/blob/master/Dockerfile#L84). On Debian system, user ids `100-999` dynamically allocate to system users and groups. See [UID and GID glasses](https://www.debian.org/doc/debian-policy/ch-opersys.html#uid-and-gid-classes). We should use not use `999` here.

**Comment by googollee** [comment]
Creating users in the image that does not exist in the host will cause many permission problems. The files created by the image have an invalid user/group as the owner, and the binary in the image can't read/write mounted files (except granting o+wr permission, which is another potential security issue) because of no such user on the host.

I propose to use `nobody:nogroup` in `Dockerfile` and guiding users to set a proper user/group when launching the image. `nobody:nogroup` is a de facto user without permission. If user doesn't provide user, it won't cause any security issues.

**Comment by kkovaletp** [comment]
Good catch! Thanks for reporting.

The fix is technically very straightforward: just replace the UID and GID from the currently used `999` to some other ID > than 1024.
Your proposal to fix might be strategically better, but we need to investigate and research it deeper to make sure that it will work for us. 
However, this is a breaking change, as after the latest release some users already configured a user (and group) on their host with the same IDs to better manage the permissions inside the volumes. So, changing this now will break their setup and force them to urgently reconfigure the user on the host and fix all previously granted permissions of the content in the volumes.

**Comment by googollee** [comment]
I hope not to create any user/group, but using `nobody:nogroup` by default, or `www-data:www-data`.

Docker doesn't isolate user id and group id between containers and host, if not [enabling user ns remapping](https://docs.docker.com/engine/security/userns-remap/). The user in the container shares the same permissions with the same id user on the host. Imaging two cases:

- User id `999` in the container and the host doesn't have any users with id `999`. 
  - The image creates caches on the host with 999 as the owner. When list those files, the owner is showed as id `999`.
  - The image can't read mounted files if they don't grant `o+r`.
    - The user can't change mounted files to the user/group with id 999, as there is no such user/group on the host.
 - User id `999` in the container and the host has another high-priority user with the same id `999`.
   - Suddenly the container runs with a high-priority user, as with the same id, the user in the container is the user on the host.
   - If photoview has any vulnerabilities, it'll be a chance to use it with more permissions.

`nobody:nogroup` and `www-data:www-data` as common users/groups across the env, they always have minimum permissions on the system. In the first case, the files have a proper user on the host. In the second case, as photoview runs with lower permissions, it's hard to use vulnerabilities.

**Comment by kkovaletp** [comment]
Good explanation. I agree with you on the most statements.

> * The image can't read mounted files if they don't grant `o+r`

We have a workaround command for that, but I agree that it is not so good solution.

> * The user can't change mounted files to the user/group with ID 999, as there is no such user/group on the host.

if those are files in the cache folder, I don't see a need for a user to modify them (or even touch them). If those are files in the source media library, they are supposed to be owned by a host user and shared with the 999 user or everyone to read. Also not so a good solution...

> * Suddenly the container runs with a high-priority user, as with the same ID, the user in the container is the user on the host.
> * If photoview has any vulnerabilities, it'll be a chance to use it with more permissions.

In theory, yes, it might be more dangerous compared to a user with ID > 1024. But in practice (as far as I understand), we create a new user and a new group in the system, so this particular UID/GID has no system permissions assigned inside the container, as they are assigned by the system at the privileged user creation time by a service setup procedure. This means that if a user was created manually even with an ID from the 1st 1024 pool, it has regular permissions and access, like any other user with ID > 1024. There is still a risk of escaping the Docker jail and using the permissions of an existing 999 system user on the host. But escaping the Docker jail is a really difficult step for malicious software, so this risk is minimal (from my PoV)

**Comment by kkovaletp** [comment]
We have to revert our base image back to the Debian, so this issue is again valid - reopenning

**Comment by googollee** [comment]
Should I change it to 9999 with `bookworm` now?

**Comment by kkovaletp** [comment]
@googollee, no, actually you didn't: https://github.com/photoview/photoview/blob/a6cbfc76f3d4acbf9dd3e5767a34ba319fc4cd56/Dockerfile#L83
And let's not implement that now, as there is no showstopper for this issue in Debian, so we can postpone it till the next major release

**Comment by kkovaletp** [comment]
@googollee, as I mentioned before, let's postpone this fix till the next major version release, as now we have an option to keep the current implementation.
If you don't see a strong need to merge it now, please just revert the Readme to the 999 values now.
Once we have some unavoidable breaking changes, we'll merge this fix too - this will let us have only 1 major version

**Comment by googollee** [comment]
Okay. As you wish. Updated #1044.

**Comment by terslowe** [comment]
What is the current status of this issue with 999 UID/GID ?

**Comment by kkovaletp** [comment]
it is in the backlog, waiting for the next major version scope.
it is not going to be part of the next release for sure.

**Comment by kkovaletp** [comment]
We don't have a release schedule. We're working now on the next release, as it goes with no release date planned because we're only 2 active contributors and nobody is working on the project full time
